src/common/transformations/src/transformations/mlir/op/binary_eltwise.cpp

```diff
@@ -0,0 +1,122 @@
+
+#include <openvino/op/relu.hpp>
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+
+#include "binary_eltwise.hpp"
+
+namespace {
+
+using namespace ov;
+using namespace ov::mlir;
+using ::mlir::ValueRange;
+
+class ConvertBinaryEltwise {
+
+    BinaryEltwisePatternBase::Builder m_op_builder;
+
+public:
+
+    ConvertBinaryEltwise(BinaryEltwisePatternBase::Builder op_builder) : m_op_builder(op_builder) {}
+
+    void operator()(ConversionContext& context, NodePtr node) {
+        auto loc = createLocation(context.context, node);
+        auto& builder = context.builder();
+        const auto inputs = context.getInputs(node);
+        const auto ov_output_element_type = node->get_output_element_type(0);
+        const auto ov_output_shape = node->get_output_partial_shape(0);
+        auto outType = importTensor(context.context, ov_output_shape, ov_output_element_type);
+        const int output_rank = ov_output_shape.rank().get_length();
+
+        SmallVector<Value> dynamicSizes;
+        for (auto [idx, dim] : llvm::enumerate(ov_output_shape)) {
+            if (!dim.is_dynamic())
+                continue;
+            dynamicSizes.push_back(context.get_dimension_value(dim));
+        }
+
+        SmallVector<Value> broadcasted_inputs;
+        for(size_t i = 0; i < inputs.size(); ++i) {
+            auto dimensions = broadcast_dimensions(node->get_input_partial_shape(i), ov_output_shape);
+            if(!dimensions.empty()) {
+                // FIXME: Find a way to avoid dimension squeezing before applying linalg.broadcast
+
+                // Step 1: Squeeze input shape to eliminate broadcasted dimensions
+                SmallVector<ReassociationIndices, 6> squeeze_map;
+                ReassociationIndices ri_cur;
+                size_t output_idx = 0; // index in ov_output_shape
+                bool group_open = true;
+                for(auto [_, dim]: llvm::enumerate(dimensions)) {
+                    for(; output_idx < dim; ++output_idx) {
+                        if(!ri_cur.empty() && !group_open) {
+                            squeeze_map.emplace_back(ri_cur);
+                            ri_cur = ReassociationIndices();
+                        }
+                        ri_cur.push_back(output_idx);
+                        group_open = false;
+                    }
+                    assert(dim == output_idx);
+                    ri_cur.push_back(dim);
+                    ++output_idx;
+                }
+                for(; output_idx < output_rank; ++output_idx) {
+                    if(group_open) {
+                        ri_cur.push_back(output_idx);
+                        squeeze_map.push_back(ri_cur);
+                        group_open = false;
+                    } else {
+                        squeeze_map.push_back({output_idx});
+                    }
+                }
+
+                auto squeezed = builder.create<tensor::CollapseShapeOp>(loc, inputs[i], squeeze_map);
```

Comment:
Discussed this way of "preparation for broadcast" with @adam-smnk.
I think this approach is not convenient because the whole reason for using named ops is simplicity. If a widely used broadcast semantics requires tricks, it is not simple to use, and too low level to be an entry point dialect in OV case.

@adam-smnk, @rengolin, please propose a better way of doing that.

Reply:
Digging a bit deeper, it is the intended approach in line with the original op design:
https://discourse.llvm.org/t/rfc-primitive-ops-add-broadcastop-to-linalg/66313

Size-1 expansion is deliberately forbidden.

Reply:
You can use generics for now, but we're discussing adding broadcast and transpose to named ops.
https://discourse.llvm.org/t/rfc-transpose-attribute-for-linalg-matmul-operations/80092/

Reply:
I don't have experience using linalg generics in C++, so cannot quickly produce this code. I will merge it as-is with collapse/broadcast.